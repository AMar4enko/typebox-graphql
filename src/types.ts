import * as T from '@sinclair/typebox/type'
import type { Simplify, UnionToIntersection } from 'type-fest'
import * as GraphQLKind from './kind.ts'

export type AddNameTag<T extends T.TProperties, Name extends string> = T & {
  _tag: T.TOptional<T.TLiteral<Name>>
}

export declare namespace GraphQL {
  namespace Utils {
    type InterfaceProps<I> = I extends Interface<infer N, infer P, infer E>
      ? P
      : // biome-ignore lint/complexity/noBannedTypes: <explanation>
        {}

    type ExtendInterface<I extends string> =
      | {
          [GraphQLKind.Interface]: I
        }
      | {
          [GraphQLKind.Object]: I
        }
  }

  interface Interface<
    Name extends string = string,
    T extends T.TProperties = T.TProperties,
    Ext extends string = string,
  > extends T.TObject<AddNameTag<T, Name>> {
    [GraphQLKind.Interface]: {
      name: Name
      extends: GraphQL.Interface[]
    }
  }

  interface Object<
    Name extends string = string,
    T extends T.TProperties = T.TProperties,
    Ext extends string = string,
  > extends T.TObject<AddNameTag<T, Name>> {
    [GraphQLKind.Object]: {
      name: Name
      extends: Ext[]
    }
  }

  interface Union<Name extends string, Elements extends Object | Interface>
    extends T.TUnion<Elements[]> {
    [GraphQLKind.Union]: {
      name: Name
      extends: Elements[]
    }
  }

  interface Enum<Name extends string, Elements extends string>
    extends T.TUnion<T.TLiteral<Elements>[]> {
    [GraphQLKind.Enum]: {
      name: Name
      values: T.TLiteral<Elements[number]>[]
    }
  }

  interface WithArgs<S extends T.TSchema, Args extends T.TProperties>
    extends T.TSchema {
    static: T.Static<S>
    [GraphQLKind.FieldWithArgs]: Args
  }

  interface ID extends T.TString {
    [GraphQLKind.ID]: GraphQLKind.Kind.ID
  }
}

const createSchema = (
  kind: GraphQLKind.Kind.Interface | GraphQLKind.Kind.Object,
) =>
  function (
    name: string,
    props: T.TProperties,
    options?: T.ObjectOptions & {
      extends?: GraphQL.Interface<string, T.TProperties>[]
    },
  ) {
    const interfacesProps = (options?.extends ?? []).reduce((acc, s) => {
      return Object.assign(acc, s.properties)
    }, {} as T.TProperties)

    return Object.assign(
      T.Object(
        Object.assign({}, props, interfacesProps, {
          _tag: T.Optional(T.Literal(name)),
        }),
        {
          ...options,
          $id: name,
        },
      ),
      {
        [kind]: { name, extends: options?.extends ?? [] },
      },
      // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    ) as any
  }

const GraphQLInterface: {
  <
    Name extends string,
    Props extends T.TProperties,
    Ext extends GraphQL.Interface<string, T.TProperties, string>,
  >(
    name: Name,
    props: Props,
    options: T.ObjectOptions & { extends: Ext[] },
  ): GraphQL.Interface<
    Name,
    [keyof Props] extends [never]
      ? Simplify<UnionToIntersection<GraphQL.Utils.InterfaceProps<Ext>>>
      : Simplify<
          UnionToIntersection<GraphQL.Utils.InterfaceProps<Ext>> & Props
        >,
    Ext[GraphQLKind.Kind.Interface]['name']
  >
  <Name extends string, Props extends T.TProperties>(
    name: Name,
    props: Props,
    options?: T.ObjectOptions,
  ): GraphQL.Interface<Name, Props, never>
} = createSchema(GraphQLKind.Interface)

export const IsInterface = (value: T.TSchema): value is GraphQL.Interface =>
  GraphQLKind.Interface in value

const GraphQLObject: {
  <
    Name extends string,
    Props extends T.TProperties,
    Ext extends GraphQL.Interface<string, T.TProperties, string>,
  >(
    name: Name,
    props: Props,
    options: T.ObjectOptions & { extends: Ext[] },
  ): GraphQL.Object<
    Name,
    [keyof Props] extends [never]
      ? Simplify<UnionToIntersection<GraphQL.Utils.InterfaceProps<Ext>>>
      : Simplify<
          UnionToIntersection<GraphQL.Utils.InterfaceProps<Ext>> & Props
        >,
    Ext[GraphQLKind.Kind.Interface]['name']
  >
  <Name extends string, Props extends T.TProperties>(
    name: Name,
    props: Props,
    options?: T.ObjectOptions,
  ): GraphQL.Object<Name, Props, never>
} = createSchema(GraphQLKind.Object)

export const IsObject = (value: T.TSchema): value is GraphQL.Object =>
  GraphQLKind.Object in value

const GraphQLUnion = <
  Name extends string,
  Elements extends GraphQL.Object | GraphQL.Interface,
>(
  name: Name,
  elements: Elements[],
  options?: T.SchemaOptions,
): GraphQL.Union<Name, Elements> =>
  Object.assign(
    T.Union(elements, {
      ...options,
      $id: name,
    }),
    {
      [GraphQLKind.Union]: { name },
    },
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  ) as any

export const IsUnion = (
  value: T.TSchema,
): value is GraphQL.Union<string, never> => GraphQLKind.Union in value

const GraphQLEnum = <Name extends string, Elements extends string>(
  name: string,
  elements: Elements[],
): GraphQL.Enum<Name, Elements> & { [key in Elements]: key } => {
  const elem = elements.map((e) => T.Literal(e))
  const union = T.Union(elem, {
    $id: name,
  })

  return Object.assign(union, {
    [GraphQLKind.Enum]: { name, values: union.elements },
    ...elements.reduce(
      (acc, e) => {
        acc[e] = e
        return acc
      },
      {} as { [key in Elements[number]]: Elements },
    ),
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  }) as any
}

export const GraphQLFieldWithArgs = <
  S extends T.TSchema,
  Args extends T.TProperties,
>(
  s: S,
  args: Args,
) => {
  return Object.assign(T.CloneType(s), {
    [GraphQLKind.FieldWithArgs]: args,
  }) as GraphQL.WithArgs<S, Args>
}

export const IsFieldWithArgs = (
  x: T.TSchema,
): x is GraphQL.WithArgs<T.TSchema, T.TProperties> => {
  return GraphQLKind.FieldWithArgs in x
}

export const ID = () => {
  return Object.assign(T.String(), {
    [GraphQLKind.ID]: GraphQLKind.ID,
  }) as GraphQL.ID
}

export const IsID = (value: T.TSchema): value is GraphQL.ID =>
  GraphQLKind.ID in value

export {
  GraphQLInterface as Interface,
  GraphQLObject as Object,
  GraphQLUnion as Union,
  GraphQLEnum as Enum,
  GraphQLFieldWithArgs as FieldWithArgs,
}
