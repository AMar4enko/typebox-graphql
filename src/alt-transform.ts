/**
 * TypeBox Transform alternative, that captures
 * both side schemas, not only input side
 */

import type {
  SchemaOptions,
  TSchema,
  TTransform,
  TransformOptions,
} from '@sinclair/typebox'
import { Clone, CloneType, Transform } from '@sinclair/typebox/type'

import type { Static, StaticDecode } from '@sinclair/typebox'

export const AltTransformKind: unique symbol = Symbol(
  '@sinclair/typebox/symbols/AltTransformKind',
)

export interface TAltTransform<
  I extends TSchema = TSchema, // I stands for Schema on the input side
  A extends TSchema = TSchema, // A stands for Schema on the output side
> extends TTransform<I, Static<A>> {
  // static: TransformStatic<A, this['params']>
  [AltTransformKind]: I
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  [key: string]: any
}

export const AltTransform = <A extends TSchema, I extends TSchema>(
  from: I,
  to: A,
  options: {
    decode: (fromA: Static<I>) => Static<A>
    encode: (toA: Static<A>) => Static<I>
  } & SchemaOptions,
): TAltTransform<I, A> => {
  const { encode, decode, ...opts } = options
  const originalTransform = Transform(from)
    .Decode(options.decode)
    .Encode(options.encode)

  return Object.assign(originalTransform, {
    [AltTransformKind]: CloneType(to, opts),
  }) as unknown as TAltTransform<I, A>
}

export const IsAltTransform = (x: TSchema): x is TAltTransform => {
  return AltTransformKind in x
}
