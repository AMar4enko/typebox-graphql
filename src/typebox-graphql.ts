import * as T from '@sinclair/typebox'

import {
  type GraphQLFieldConfig,
  GraphQLObjectType,
  type GraphQLResolveInfo,
  GraphQLSchema,
} from 'graphql'
import { Compiler } from './compiler/index.ts'
import * as Kind from './kind.ts'
import { type GraphQL, IsFieldWithArgs } from './types.ts'

type ReadonlyOptionalPropertyKeys<T extends T.TProperties> = {
  [K in keyof T]: T[K] extends T.TReadonly<T.TSchema>
    ? T[K] extends T.TOptional<T[K]>
      ? K
      : never
    : never
}[keyof T]
type ReadonlyPropertyKeys<T extends T.TProperties> = {
  [K in keyof T]: T[K] extends T.TReadonly<T.TSchema>
    ? T[K] extends T.TOptional<T[K]>
      ? never
      : K
    : never
}[keyof T]
type OptionalPropertyKeys<T extends T.TProperties> = {
  [K in keyof T]: T[K] extends T.TOptional<T.TSchema>
    ? T[K] extends T.TReadonly<T[K]>
      ? never
      : K
    : never
}[keyof T]
type RequiredPropertyKeys<T extends T.TProperties> = keyof Omit<
  T,
  | ReadonlyOptionalPropertyKeys<T>
  | ReadonlyPropertyKeys<T>
  | OptionalPropertyKeys<T>
>

// prettier-ignore
type ObjectStaticProperties<
  T extends T.TProperties,
  R extends Record<PropertyKey, unknown>,
> = T.Evaluate<
  Readonly<Partial<Pick<R, ReadonlyOptionalPropertyKeys<T>>>> &
    Readonly<Pick<R, ReadonlyPropertyKeys<T>>> &
    Partial<Pick<R, OptionalPropertyKeys<T>>> &
    Required<Pick<R, RequiredPropertyKeys<T>>>
>

export declare namespace SchemaRegistry {
  type ResolvableType = GraphQL.Interface | GraphQL.Object

  type IsInterface<A> = A extends GraphQL.Interface
    ? A
    : A extends T.TArray<infer MaybeI extends GraphQL.Interface>
      ? MaybeI
      : never

  type AsString<T> = T extends string ? T : never

  type SafeKey<A> = [A] extends [never]
    ? never
    : keyof A extends string
      ? keyof A
      : never

  type TypeName<T extends ResolvableType> = T extends GraphQL.Interface<
    infer Name
  >
    ? Name
    : T extends GraphQL.Object<infer Name>
      ? Name
      : never

  /**
   * Creates union of all resolvable type properties
   * in <Name>.<PropertyName> format
   */
  type ResolvableTypeProperties<T extends ResolvableType> = T extends {
    properties: T.TProperties
  }
    ? `${AsString<TypeName<T>>}.${AsString<Exclude<keyof T['properties'], '_tag'>>}`
    : never

  type ResolverKeys<R> = R extends SchemaRegistry<
    infer Type,
    infer Query,
    infer Mutation,
    infer Subscription,
    infer Resolver
  >
    ?
        | ResolvableTypeProperties<Type[keyof Type]>
        | `Query.${Query extends T.TObject<T.TProperties> ? SafeKey<Query['properties']> : never}`
        | `Mutation.${Mutation extends T.TObject<T.TProperties> ? SafeKey<Mutation['properties']> : never}`
        | `Subscription.${Subscription extends T.TObject<T.TProperties> ? SafeKey<Subscription['properties']> : never}`
    : never

  /**
   * Creates union of all schema fields that are not yet resolved
   */
  type UnresolvedKey<R> = R extends SchemaRegistry<
    infer Type,
    infer Query,
    infer Mutation,
    infer Subscription,
    infer Resolver
  >
    ? Exclude<
        ResolverKeys<R>,
        [Resolver] extends [never] ? never : keyof Resolver
      >
    : never

  export type ObjectsExtendingInterface<
    Types extends ResolvableType,
    Needle extends string,
  > = Types extends GraphQL.Object<
    infer Name,
    infer Props,
    infer InterfaceNames
  >
    ? Needle extends InterfaceNames
      ? Types
      : never
    : never

  export type InterfacesExtendingInterface<
    Types extends ResolvableType,
    Needle extends string,
  > = Types extends GraphQL.Interface<
    infer Name,
    infer Props,
    infer InterfaceNames
  >
    ? Needle extends InterfaceNames
      ? Types
      : never
    : never

  export type GetInterfaceNames<Name extends string> =
    | GraphQL.Interface<Name, T.TProperties, string>
    | GraphQL.Object<Name, T.TProperties, string>

  export type TypesExtendingInterface<
    Types extends ResolvableType,
    Needle extends string,
    Acc = never,
  > =
    | Acc
    | ObjectsExtendingInterface<Types, Needle>
    | InterfacesExtendingInterface<Types, Needle>

  /**
   * Based on the registry state returns all types
   * extending a specific interface. Limited to 5 levels deep
   */
  type AllTypesExtendingInterface<
    RegisteredTypes extends ResolvableType,
    I extends string,
  > = TypesExtendingInterface<RegisteredTypes, I> extends infer Types
    ? Types extends GetInterfaceNames<infer Names>
      ? TypesExtendingInterface<
          RegisteredTypes,
          Names,
          Types
        > extends infer Types
        ? Types extends GetInterfaceNames<infer Names>
          ? TypesExtendingInterface<
              RegisteredTypes,
              Names,
              Types
            > extends infer Types
            ? Types extends GetInterfaceNames<infer Names>
              ? TypesExtendingInterface<
                  RegisteredTypes,
                  Names,
                  Types
                > extends infer Types
                ? Types extends GetInterfaceNames<infer Names>
                  ? TypesExtendingInterface<RegisteredTypes, Names, Types>
                  : never
                : never
              : never
            : never
          : never
        : never
      : never
    : never

  /**
   * Code below generates resolver function result type from schema
   * and registered types, deep expanding interfaces to possible arbitrary object
   * alternatives
   */

  /**
   * When expanding object we first check if any of it's properties
   * is GraphQL interface type, short circuiting to TypeBox Static if none
   */
  type ExpandObject<
    S,
    RegisteredTypes extends ResolvableType,
  > = S extends GraphQL.Object<infer Name, infer P, infer Extends>
    ? [IsInterface<S['properties'][keyof S['properties']]>] extends [never]
      ? T.Static<S>
      : ObjectStaticProperties<
          S['properties'],
          {
            [key in keyof S['properties']]: ExpandedResultType<
              S['properties'][key],
              RegisteredTypes
            >
          }
        >
    : never

  type ExpandInterface<
    S,
    RegisteredTypes extends ResolvableType,
  > = S extends GraphQL.Interface<infer Name, infer P, infer Extends>
    ?
        | ([IsInterface<S['properties'][keyof S['properties']]>] extends [never]
            ? T.Static<S>
            : ObjectStaticProperties<
                S['properties'],
                {
                  [key in keyof S['properties']]: ExpandedResultType<
                    S['properties'][key],
                    RegisteredTypes
                  >
                }
              >)
        | ExpandedResultType<
            AllTypesExtendingInterface<RegisteredTypes, Name>,
            RegisteredTypes
          >
    : never

  type ExpandArraySchema<
    S,
    RegisteredTypes extends ResolvableType,
  > = S extends T.TArray<infer E>
    ? ExpandedResultType<E, RegisteredTypes>[]
    : never

  type ExpandUnionSchema<
    S,
    RegisteredTypes extends ResolvableType,
  > = S extends T.TUnion<infer E>
    ? ExpandedResultType<E[number], RegisteredTypes>
    : never

  type ExpandedResultType<
    S extends T.TSchema,
    RegisteredTypes extends ResolvableType,
  > =
    | ExpandObject<S, RegisteredTypes>
    | ExpandInterface<S, RegisteredTypes>
    | ExpandUnionSchema<S, RegisteredTypes>
    | ExpandArraySchema<S, RegisteredTypes> extends infer A
    ? [A] extends [never]
      ? T.Static<S>
      : A
    : never

  type FieldArgs<S> = S extends GraphQL.WithArgs<
    infer _,
    infer Args extends T.TProperties
  >
    ? T.Static<T.TObject<Args>>
    : never

  type ResolverFunction<
    S extends T.TSchema,
    RegisteredTypes extends ResolvableType,
    Ctx = never,
    Root extends T.TObject = never,
  > = (
    root: [Root] extends [never] ? never : T.Static<Root>,
    args: FieldArgs<S>,
    ctx: Ctx,
    info: GraphQLResolveInfo,
  ) =>
    | ExpandedResultType<S, RegisteredTypes>
    | Promise<ExpandedResultType<S, RegisteredTypes>>

  type SchemaResolverFunction<R, Key extends UnresolvedKey<R>> = [R] extends [
    SchemaRegistry<
      infer Type,
      infer Query,
      infer Mutation,
      infer _Subscription,
      infer _Resolver,
      infer Ctx
    >,
  ]
    ? Key extends `${infer TypeName}.${infer FieldName}`
      ? [TypeName, FieldName] extends [
          SafeKey<Type>,
          SafeKey<Type[TypeName]['properties']>,
        ]
        ? ResolverFunction<
            Type[TypeName][FieldName],
            Type[keyof Type],
            Ctx,
            Type[TypeName]
          >
        : [TypeName, FieldName] extends [`Query`, SafeKey<Query['properties']>]
          ? ResolverFunction<
              Query['properties'][FieldName],
              Type[keyof Type],
              Ctx
            >
          : [TypeName, FieldName] extends [
                `Mutation`,
                SafeKey<Mutation['properties']>,
              ]
            ? ResolverFunction<
                Mutation['properties'][FieldName],
                Type[keyof Type],
                Ctx
              >
            : never
      : never
    : never
}

export interface SchemaRegistry<
  Type extends { [key in string]: GraphQL.Object } = never,
  Query extends T.TObject<{
    [name in string]: T.TSchema
  }> = never,
  Mutation extends T.TObject<{
    [name in string]: T.TSchema
  }> = never,
  Subscription extends T.TObject<{
    [name in string]: T.TSchema
  }> = never,
  Resolver extends { [key in string]: unknown } = never,
  // Unresolved extends { [key in string]: SchemaRegistry.CachedUnresolved } = never,
  Ctx = never,
> {
  readonly type: Map<string, GraphQL.Object>
  readonly query: Query
  readonly mutation: Mutation
  readonly subscription: Map<
    keyof Subscription,
    Subscription[keyof Subscription]
  >
  readonly resolver: Map<keyof Resolver, Resolver[keyof Resolver]>

  compile: () => GraphQLSchema

  addType: <S extends GraphQL.Object>(
    type: S,
  ) => SchemaRegistry<
    [Type] extends [never]
      ? { [key in SchemaRegistry.TypeName<S>]: S }
      : { [key in SchemaRegistry.TypeName<S>]: S } & Type,
    Query,
    Mutation,
    Subscription,
    Resolver,
    Ctx
  >

  setQuery: <S extends T.TObject>(
    query: S,
  ) => SchemaRegistry<Type, S, Mutation, Subscription, Resolver, Ctx>

  setMutation: <S extends T.TObject>(
    mutation: S,
  ) => SchemaRegistry<Type, Query, S, Subscription, Resolver, Ctx>

  setSubscription: <S extends T.TObject>(
    subscription: S,
  ) => SchemaRegistry<Type, Query, Mutation, S, Resolver, Ctx>

  resolve: <
    Key extends SchemaRegistry.UnresolvedKey<
      SchemaRegistry<Type, Query, Mutation, Subscription, Resolver>
    >,
  >(
    key: Key,
    resolverFn: SchemaRegistry.SchemaResolverFunction<
      SchemaRegistry<Type, Query, Mutation, Subscription, Resolver>,
      Key
    >,
  ) => SchemaRegistry<
    Type,
    Query,
    Mutation,
    Subscription,
    Resolver & { [id in Key]: typeof resolverFn },
    Ctx
  >
}

/**
 * @internal
 */

function addType(this: SchemaRegistry, type: T.TSchema) {
  return createProto({
    ...this,
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    type: new Map([...this.type, [type.$id, type]]) as any,
  })
}

/**
 * @internal
 */

function withOperation(operationType: `mutation` | `query` | `subscription`) {
  return function withOperation(this: SchemaRegistry, type: T.TObject) {
    return createProto({
      ...this,
      [operationType]: type,
    })
  }
}

function resolve(this: SchemaRegistry, key: string, resolverFn: unknown) {
  return createProto({
    ...this,
    resolver: new Map([...this.resolver, [key, resolverFn]]),
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  } as any)
}

function createProto(r: SchemaRegistry) {
  return Object.freeze(
    Object.assign(
      Object.create({
        addType,
        setQuery: withOperation(`query`),
        setMutation: withOperation(`mutation`),
        resolve,
        compile,
      }),
      r,
    ),
  )
}

export const empty = (): SchemaRegistry<never, never, never, never, never> => {
  const obj = createProto({
    type: new Map(),
    query: new Map(),
    mutation: new Map(),
    subscription: new Map(),
    resolver: new Map(),
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  } as any)

  Object.freeze(obj)

  return obj
}

// biome-ignore lint/suspicious/noExplicitAny: <explanation>
function compile(this: SchemaRegistry<any, any, any, any, any, any>) {
  const c = new Compiler()
  this.type.values().map((a: GraphQL.Object) => c.compileOutputType(a))
  const queryObj: T.TObject<{ [key in string]: GraphQL.Object }> =
    this.query ?? T.Object({})

  const query = [
    ...Object.entries(queryObj.properties).map(
      ([key, type]: [string, GraphQL.Object]) => {
        return [
          key,
          {
            type: c.compileAsOutputType(type),
            args: IsFieldWithArgs(type)
              ? // @ts-ignore
                c.compilePropertiesAsArgs(type[Kind.FieldWithArgs])
              : {},
            description: type.description,
            resolve: this.resolver.get(`Query.${String(key)}`),
            // biome-ignore lint/suspicious/noExplicitAny: <explanation>
          } satisfies GraphQLFieldConfig<any, any, any>,
        ]
      },
    ),
  ]

  const mutationObj: T.TObject<{ [key in string]: GraphQL.Object }> =
    this.mutation ?? T.Object({})

  const mutation = [
    ...Object.entries(mutationObj.properties).map(([key, type]) => {
      return [
        key,
        {
          type: c.compileAsOutputType(type),
          args: IsFieldWithArgs(type)
            ? c.compilePropertiesAsArgs(type[Kind.FieldWithArgs])
            : {},
          description: type.description,
          resolve: this.resolver.get(`Mutation.${String(key)}`),
          // biome-ignore lint/suspicious/noExplicitAny: <explanation>
        } satisfies GraphQLFieldConfig<any, any, any>,
      ]
    }),
  ]
  const types = c.getNamedTypes()
  return new GraphQLSchema({
    types,
    query:
      query.length > 0
        ? new GraphQLObjectType({
            name: `Query`,
            fields: Object.fromEntries([...query]),
          })
        : undefined,
    mutation:
      mutation.length > 0
        ? new GraphQLObjectType({
            name: `Mutation`,
            fields: Object.fromEntries([...mutation]),
          })
        : undefined,
  })
}
