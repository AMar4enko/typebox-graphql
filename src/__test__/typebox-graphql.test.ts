import { describe, expect, test } from 'bun:test'
import { useGraphQLSSE } from '@graphql-yoga/plugin-graphql-sse'
import * as T from '@sinclair/typebox/type'
import { serve } from 'bun'
import { printSchema } from 'graphql'
import { createClient } from 'graphql-sse'
import { type YogaServerInstance, createYoga } from 'graphql-yoga'
import { Hono } from 'hono'
import { createFactory } from 'hono/factory'
import type { ContentfulStatusCode } from 'hono/utils/http-status'
import { AltTransform } from '../alt-transform.ts'
import { empty } from '../typebox-graphql.ts'
import { FieldWithArgs, ID, Interface, Object } from '../types.ts'

export const appFactory = createFactory()

// biome-ignore lint/complexity/noBannedTypes: <explanation>
const createYogaMiddleware = (yoga: YogaServerInstance<{}, {}>) =>
  appFactory.createHandlers(async (c) => {
    const { body, status, headers } = await yoga.fetch(c.req.raw, c)

    if (!body) {
      return c.body(null, 204)
    }

    return c.body(body, {
      status: status as ContentfulStatusCode,
      headers,
    })
  })

const DateFromString = AltTransform(T.String(), T.Date(), {
  decode: (s: string) => new Date(s),
  encode: (d: Date) => d.toISOString(),
  $id: `DateFromString`,
})

type A = T.Static<typeof DateFromString>

const Likeable = Interface(`Likeable`, {
  likes: T.Integer(),
})

const User = Object(
  `User`,
  {
    id: ID(),
    createdAt: DateFromString,
    email: T.String(),
    subscribers: T.Array(T.String()),
  },
  { extends: [Likeable] },
)

const UpdateUserPayload = T.Object(
  {
    email: T.String(),
  },
  { $id: `UpdateUserPayload` },
)

const Query = T.Object({
  getUser: FieldWithArgs(User, { id: ID() }),
})

const Mutation = T.Object({
  updateUser: FieldWithArgs(User, { id: ID(), payload: UpdateUserPayload }),
})

const Subscription = T.Object({
  userUpdates: FieldWithArgs(User, { id: ID() }),
})

const reg = empty()
  .addType(User)
  .setQuery(Query)
  .setMutation(Mutation)
  .setSubscription(Subscription)

describe(`Schema compilation`, () => {
  test(`it works`, () => {
    const schema = reg.compile()
    expect(printSchema(schema)).toMatchSnapshot()
  })
})

describe(`E2E`, () => {
  test(
    `works`,
    async () => {
      const schema = reg
        .resolve(`Query.getUser`, (root, args) => {
          return {
            _tag: `User`,
            id: args.id,
            createdAt: new Date(),
            email: `whatever@user.com`,
            likes: 0,
            subscribers: [],
          }
        })
        .compile()

      const yoga = createYoga({
        landingPage: true,
        schema,
        plugins: [useGraphQLSSE()],
      })

      const app = new Hono().all(
        `/graphql/stream`,
        ...createYogaMiddleware(yoga),
      )

      const server = serve({
        fetch(request, server) {
          if (request.headers.get('accept') === 'text/event-stream') {
            server.timeout(request, 3600)
          }

          return app.fetch(request)
        },
        port: 3000,
      })

      const client = createClient({
        url: 'http://localhost:3000/graphql/stream',
      })

      const resp = await client
        .iterate({
          query: `query { getUser(id: "1") { id createdAt email subscribers likes } }`,
        })
        .next()

      expect(resp.value.data.getUser).toMatchObject({
        id: expect.anything(),
        createdAt: new Date(),
        email: `whatever@user.com`,
        likes: 0,
        subscribers: [],
      })

      server.stop()
    },
    { timeout: 5_000 },
  )
})
