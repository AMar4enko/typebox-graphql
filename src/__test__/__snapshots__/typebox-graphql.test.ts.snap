// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema compilation it works 1`] = `
"input UpdateUserPayload {
  email: String!
}

type User implements Likeable {
  id: ID!
  createdAt: DateFromString!
  email: String!
  subscribers: [String!]!
  likes: Int!
}

scalar DateFromString

interface Likeable {
  likes: Int!
}

type Query {
  getUser(id: ID!): User!
}

type Mutation {
  updateUser(id: ID!, payload: UpdateUserPayload!): User!
}

type Subscription {
  userUpdates(id: ID!): User!
}"
`;
